pkg/console/subresource/configmap/configmap_builder.go

```diff
@@ -0,0 +1,156 @@
+// return the default config value immediately:
+//   b.Config()
+// set all the values:
+//   b.SetHost(host).SetLogoutURL().setBrand().setDocURL().SetAPIServerURL().SetProductName().SetCustomLogoFile().Config()
+// set only some values:
+//   b.SetHost().SetBrand().Config()
+// unit test each permutation of the builder to be confident it does what it is supposed to do without breaking!
+type YamlConfigBuilder interface {
+	Host(host string) *YamlConfigBuilder
+	LogoutURL(logoutRedirectURL string) *YamlConfigBuilder
+	Brand(brand operatorv1.Brand) *YamlConfigBuilder
+	DocURL(docURL string) *YamlConfigBuilder
+	APIServerURL(apiServerURL string) *YamlConfigBuilder
+	CustomProductName(productName string) *YamlConfigBuilder
+	CustomLogo(customLogoFile string) *YamlConfigBuilder
+	StatusPageID(id string) *YamlConfigBuilder
+	// the config as structs
+	Config() (consoleConfig yaml.MapSlice, marshallError error)
+}
+
+type ConsoleYAMLConfigBuilder struct {
+	host              string
+	logoutRedirectURL string
+	brand             operatorv1.Brand
+	docURL            string
+	apiServerURL      string
+	statusPageID      string
+	// TODO: pipe these in when they added
+	customProductName string
+	customLogoFile    string
+}
+
+func (b *ConsoleYAMLConfigBuilder) Host(host string) *ConsoleYAMLConfigBuilder {
+	b.host = host
+	return b
+}
+func (b *ConsoleYAMLConfigBuilder) LogoutURL(logoutRedirectURL string) *ConsoleYAMLConfigBuilder {
+	b.logoutRedirectURL = logoutRedirectURL
+	return b
+}
+func (b *ConsoleYAMLConfigBuilder) Brand(brand operatorv1.Brand) *ConsoleYAMLConfigBuilder {
+	b.brand = brand
+	return b
+}
+func (b *ConsoleYAMLConfigBuilder) DocURL(docURL string) *ConsoleYAMLConfigBuilder {
+	b.docURL = docURL
+	return b
+}
+func (b *ConsoleYAMLConfigBuilder) APIServerURL(apiServerURL string) *ConsoleYAMLConfigBuilder {
+	b.apiServerURL = apiServerURL
+	return b
+}
+
+func (b *ConsoleYAMLConfigBuilder) StatusPageID(id string) *ConsoleYAMLConfigBuilder {
+	b.statusPageID = id
+	return b
+}
+
+func (b *ConsoleYAMLConfigBuilder) Config() yaml.MapSlice {
+	conf := yaml.MapSlice{}
+	// TODO: defaulting should be done here
+	// NO global default strings & such should be needed, everything hidden in this func.
+	conf = append(conf, yaml.MapItem{Key: "kind", Value: "ConsoleConfig"})
+	conf = append(conf, yaml.MapItem{Key: "apiVersion", Value: "console.openshift.io/v1"})
+	conf = append(conf, yaml.MapItem{Key: "auth", Value: b.authServer(b.logoutRedirectURL)})
+	conf = append(conf, yaml.MapItem{Key: "clusterInfo", Value: b.clusterInfo(b.host, b.apiServerURL)})
+	conf = append(conf, yaml.MapItem{Key: "customization", Value: b.customization(b.brand, b.docURL)})
+	conf = append(conf, yaml.MapItem{Key: "servingInfo", Value: b.servingInfo()})
+	conf = append(conf, yaml.MapItem{Key: "providers", Value: b.providers(b.statusPageID)})
+	return conf
+}
+
+func (b *ConsoleYAMLConfigBuilder) ConfigBytes() (consoleConfigYAML []byte, marshallError error) {
```

Comment:
Not sure about the name here, maybe `ConfigToBytes()` ? (speaking as a not native person)...
Also we can probably drop named return values.

Reply:
`ConfigYAML()`  is better than `ConfigToBytes()` yeah?

I kept the named return values only when it matters: `consoleConfigYAML []byte, marshallError error` is helpful so you know `[]byte` is YAML, otherwise the function signature isn't very clear.